### PR TITLE
[skipruntime-ts] Sliced instead of MapOptions

### PR DESCRIPTION
**Pull request by mbouaziz**
As we discussed today, instead of the `MapOptions` thing introduced in #311 for ranges, we add a new function `sliced` on eager collections to do the actual slicing before mapping.
~~But it was actually simpler to implement a `sliced` that just prepares the ranges for the next `map`/`mapReduce`, saving the need to do the intermediate slicing before mapping.
Current limitation (that can be lifted later if there is a need): you cannot slice an already sliced collection. This is ensured by the type system.~~

### CODE REVIEW

**Comment by bennostein** [comment]
This makes sense, but I am not sure that it is actually _simpler_ from the user perspective -- we've introduced multiple variants of EagerCollection and TableCollection, with overlapping but not identical supported operations.

If I understand correctly, the reason for the no-slicing-a-sliced-collection restriction is to avoid computing the intersection of two sets of ranges, nothing fundamental, no problem.  But not being able to "get" without mapping is because there is no actual materialized sliced collection; the sliced collection just holds onto some ranges to be used whenever its mapped over.  Is that right?

I'd argue that this is premature optimization wrt what we talked about earlier, and that it is simpler for "slice" to apply a no-op mapper with the given ranges, producing a regular EagerCollection which supports all the usual operations.  If we want to avoid materializing those intermediate results, that seems a job for the abstract transducer-y representation that we tallked about earlier.

Am I understanding right? What do you think?

**Comment by mbouaziz** [comment]
Right.

The implementation was easier because there was no need to write a no-op map and the rangesOpt argument was already there. I agree it clutters up the public API.
I'll rewrite as an eager slicing

**Comment by mbouaziz** [comment]
Rewritten version. The hard part was getting a name for the slice

File: skipruntime-ts/ts/src/internals/skipruntime_module.ts
Comment (bennostein) [inline skipruntime-ts/ts/src/internals/skipruntime_module.ts:388]: Pros and cons of generating names this way (b64-encoding input names and joining those) vs. how I did it for merge (stripping slashes off of input names and concatenating) -- these will end up more compact on the one hand but on the other hand less human-readable.  I don't feel too strongly either way but worth keeping in mind and potentially unifying the two in the future.

Comment (mbouaziz) [reply skipruntime-ts/ts/src/internals/skipruntime_module.ts:388]: I wouldn't have done differently if the name was generated on the skip side.
Currently SKStore directory names must be `'/'`-separated paths, each path element must be either an integer or start with a letter or underscore followed by alpha-numerical chars or `'-'` and `'_'`.
But my input is any valid JSON value.

Comment (mbouaziz) [reply skipruntime-ts/ts/src/internals/skipruntime_module.ts:388]: Wait... does the name really has to be unique?

**Review by bennostein — APPROVED** [review]
LGTM!  I'll rebase on here once you merge and reuse the same identity map function over on #322

File: skipruntime-ts/ts/src/internals/skipruntime_module.ts
Comment (mbouaziz) [reply skipruntime-ts/ts/src/internals/skipruntime_module.ts:388]: I'll merge as is so we move forward but let's simplify later if we can

Comment (bennostein) [reply skipruntime-ts/ts/src/internals/skipruntime_module.ts:388]: Well, the resulting DirName/handleID does need to be unique.  But what you have here achieves that, IIUC: the "name" passed to `SkipRuntime_slice` is used as a sub-directory name (at Utils.sk line 45) on the base eager handle, so the resulting DirName is unique for each base eager collection / ranges combination

Comment (mbouaziz) [reply skipruntime-ts/ts/src/internals/skipruntime_module.ts:388]: It looks conflict-free but, to be fully accurate and fully nit picking :stuck_out_tongue_winking_eye:, it isn't because I use for delimiters characters that can be found in values. We should probably allow more characters in dirnames. Or I could have encoded the length before each value.

Comment (jberdine) [reply skipruntime-ts/ts/src/internals/skipruntime_module.ts:388]: My understanding is that the underlying SKStore Context's dirs field maps DirNames to Dirs, and everything gets ahold of a Dir through that map. So if a Dir has multiple DirNames, that is ok though perhaps confusing, but no two distinct Dirs can have the same DirName or all hell breaks loose. Right? And currently the ts layer is getting uniqueness of DirNames by mostly ad hoc means. Would it make sense to have a gensym sort of operation on the context that would e.g. generate and suffix a number onto the desired name in order to ensure it is unique with respect to the dirs in the context? If we ever make a mistake and there is a name clash, it's going to be hell for us and users to debug.

File: skipruntime-ts/src/Extern.sk
Comment (jberdine) [inline skipruntime-ts/src/Extern.sk:467]: Sorry to come late and nitpick names, but the other operation names like map and reduce are all verbs, while this one is a noun...

**Pull request merged**
